### PR TITLE
Add reading & writing of flags

### LINKED ISSUE #7: Dalli Compatibility
Hello again!

Memcachex is not compatible with Dalli, even if both of them are using JSON as the serializer.  The reason is that Dalli will only run the deserialize if a certain bit is set on the flags of the key.  Here's the relevant code:

Only deseralize if bit is set:
https://github.com/petergoldstein/dalli/blob/master/lib/dalli/server.rb#L442

Set serialization bit if a serializer is used:
https://github.com/petergoldstein/dalli/blob/master/lib/dalli/server.rb#L435

I'm not entirely sure, but I think the fix would be to set that bit in the flags on the `set`, `add`, and `replace` commands (if a coder is specified).

What do you think?

### PR DESCRIPTION
**Pull request by aspett**
Adds reading and writing of flags to achieve compatibility with Dalli (ruby client). Fixes #7. Takes inspiration from #8 and expands on comments.

- Allows passing a `flags` option to `Memcache.xxx` functions that take options.
- Supports `:serialised`, `:compressed` flags (0x1, 0x2 respectively)
- Adds a `encode_flags` callback for `Coder`s with a default passthrough impl.
- Adds a default `:serialised` flag to the `JSON` `Coder`.

There are some slightly breaking changes here, but the core API (`Memcache` module) is not affected.

1. Accessing `Memcache.Connection.execute/4` and `Memcache.Connection.execute_quiet/2` directly will give results back in a different format, potentially; `{:ok, value :: any, flags :: keyword}` instead of `{:ok, value :: any}`. I'm pretty sure it's impossible to add flag support without this change, unfortunately.
2. `Coder`s should now `use Memcache.Coder` rather than `@behaviour use Memcache.Coder` and optionally implement `encode_flags/2`.

### CODE REVIEW

**Comment by aspett** [comment]
:eyes: Tests passed, but builds failed?

**Comment by ananthakumaran** [comment]
```
$ ! mix help format > /dev/null || mix format --check-formatted
** (Mix) mix format failed due to --check-formatted.
The following files were not formatted:
  * test/memcache/connection_test.exs
  * lib/memcache/receiver.ex
  * lib/memcache/connection.ex
  * lib/memcache.ex
The command "! mix help format > /dev/null || mix format --check-formatted" exited with 1.
```

**Comment by aspett** [comment]
Thanks @ananthakumaran. I didn't notice it was running that. Have `mix format`'d it with elixir 1.7.

**Comment by aspett** [comment]
Bump

**Comment by aspett** [comment]
Bump @ananthakumaran

**Comment by ananthakumaran** [comment]
sorry for the delay. I am kind of reluctant about merging patches that add the ability to read/write flags. There are two aspects, 1) I haven't given much thought to flags during the initial implementation which makes it hard to retrofit. 2) There aren't many use cases which require the flags. I don't want a situation where you finish the PR, but I decide not to merge because of the above reasons.

With that said, here are my thoughts.

I don't like to attach meaning to flag (aka bit 1 is serialized, bit 2 is compressed, etc). As the flag page indicates this is different across language bindings. What I would instead like is, just expose it via the coder.

```elixir
encode(any, options :: Keyword.t()) :: iodata
encode(any, options :: Keyword.t()) :: {flag :: binary, iodata}
```

```elixir
@callback decode(iodata, options :: Keyword.t()) :: any
@callback decode(iodata, flag :: binary, options :: Keyword.t()) :: any
```

basically, encode can return `flag` value and decode can accept `flag`,
we probably have to make some changes to support the old coder
interface. With this, a custom coder could be used to support dalli without changes anywhere else in the application.


I am torn about the changes in `Memcache.Connection`. I don't want to break the API, but I don't have any better suggestion either. If I had used a map/struct for the return value, it would have made things a lot easier.

**Comment by wizardone** [comment]
@ananthakumaran It seems that you are not eager to merge this PR as it is. Are you willing to reconsider it or would you like to see a different approach being taken here so we can eventually have the flags functionality in place?

**Comment by ananthakumaran** [comment]
Could you explain your use case? I am not very eager at this point, but
that would change if more people want this

**Comment by aspett** [comment]
Closing due to no movement.
